src/main/java/org/terasology/metalrenegades/ai/system/AgentSpawnSystem.java

```diff
@@ -0,0 +1,113 @@
+/*
+ * Copyright 2018 MovingBlocks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.terasology.metalrenegades.ai.system;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.terasology.entitySystem.entity.EntityBuilder;
+import org.terasology.entitySystem.entity.EntityManager;
+import org.terasology.entitySystem.entity.EntityRef;
+import org.terasology.entitySystem.prefab.Prefab;
+import org.terasology.entitySystem.prefab.PrefabManager;
+import org.terasology.entitySystem.systems.BaseComponentSystem;
+import org.terasology.entitySystem.systems.RegisterMode;
+import org.terasology.entitySystem.systems.RegisterSystem;
+import org.terasology.entitySystem.systems.UpdateSubscriberSystem;
+import org.terasology.logic.location.LocationComponent;
+import org.terasology.metalrenegades.ai.component.AgentComponent;
+import org.terasology.metalrenegades.ai.component.HomeComponent;
+import org.terasology.metalrenegades.ai.component.PotentialHomeComponent;
+import org.terasology.registry.In;
+
+import java.util.Collection;
+
+/**
+ * Spawns new agents inside of available buildings with {@link PotentialHomeComponent}.
+ */
+@RegisterSystem(value = RegisterMode.AUTHORITY)
+public class AgentSpawnSystem extends BaseComponentSystem implements UpdateSubscriberSystem {
+
+    private static final Logger logger = LoggerFactory.getLogger(AgentSpawnSystem.class);
+
+    @In
+    private EntityManager entityManager;
+
+    @In
+    private PrefabManager prefabManager;
+
+    @Override
+    public void update(float delta) {
+        for (EntityRef entity : entityManager.getEntitiesWith(PotentialHomeComponent.class)) {
```

Comment:
Beware of using `update` methods that are un-throttled (literally run once every single game loop ticket), especially if they do something potentially expensive like a `entityManager.getEntitiesWith` lookup. That'll add up over time and there are usually more efficient alternatives :-)

One easy option is to throttle, that is simply add a time period during which you'll skip processing. Keep the last timestamp of when the logic ran then do a comparison vs current time and only execute if more than `x` seconds have passed. You _probably_ don't actually need to spawn agents 60 times per second :-)

Reply:
But what if I want to spawn my agents 60 times per second :)

I changed this to be on a throttle instead, set to 30 seconds between each cycle.